aws/resource_aws_cur_report_definition.go

```diff
@@ -0,0 +1,189 @@
+				Required: true,
+				ForceNew: true,
+			},
+			"s3_prefix": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(0, 256),
+			},
+			"s3_region": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"additional_artifacts": {
+				Type: schema.TypeSet,
+				Elem: &schema.Schema{Type: schema.TypeString,
+					ValidateFunc: validation.StringInSlice([]string{
+						costandusagereportservice.AdditionalArtifactQuicksight,
+						costandusagereportservice.AdditionalArtifactRedshift,
+					}, false),
+				},
+				Set:      schema.HashString,
+				Optional: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceAwsCurReportDefinitionCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).costandusagereportconn
+
+	reportName := *aws.String(d.Get("report_name").(string))
+
+	reportDefinition := &costandusagereportservice.ReportDefinition{
+		ReportName:  &reportName,
+		TimeUnit:    aws.String(d.Get("time_unit").(string)),
+		Format:      aws.String(d.Get("format").(string)),
+		Compression: aws.String(d.Get("compression").(string)),
+		S3Bucket:    aws.String(d.Get("s3_bucket").(string)),
+		S3Prefix:    aws.String(d.Get("s3_prefix").(string)),
+		S3Region:    aws.String(d.Get("s3_region").(string)),
+	}
+
+	additionalSchemaElementsSet := d.Get("additional_schema_elements").(*schema.Set)
+	for _, additionalSchemaElement := range additionalSchemaElementsSet.List() {
+		reportDefinition.AdditionalSchemaElements = append(reportDefinition.AdditionalSchemaElements,
+			aws.String(additionalSchemaElement.(string)))
+	}
+
+	additionalArtifactsSet := d.Get("additional_artifacts").(*schema.Set)
+	for _, additionalArtifact := range additionalArtifactsSet.List() {
+		reportDefinition.AdditionalArtifacts = append(reportDefinition.AdditionalArtifacts,
+			aws.String(additionalArtifact.(string)))
+	}
+
+	reportDefinitionInput := &costandusagereportservice.PutReportDefinitionInput{
+		ReportDefinition: reportDefinition,
+	}
+	log.Printf("[DEBUG] Creating AWS Cost and Usage Report Definition : %v", reportDefinitionInput)
+
+	_, err := conn.PutReportDefinition(reportDefinitionInput)
+	if err != nil {
+		return fmt.Errorf("Error creating AWS Cost And Usage Report Definition: %s", err)
+	}
+	d.SetId(reportName)
+	return resourceAwsCurReportDefinitionRead(d, meta)
+}
+
+func resourceAwsCurReportDefinitionRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).costandusagereportconn
+
+	reportName := *aws.String(d.Id())
+
+	params := &costandusagereportservice.DescribeReportDefinitionsInput{}
+
+	log.Printf("[DEBUG] Reading CurReportDefinition: %s", reportName)
+
+	var matchingReportDefinition *costandusagereportservice.ReportDefinition
+	err := conn.DescribeReportDefinitionsPages(params, func(resp *costandusagereportservice.DescribeReportDefinitionsOutput, isLast bool) bool {
```

Comment:
Since we repeat this logic a few times, it seems like it would be helpful to wrap this in a function, e.g.

```go
func describeCurReportDefinition(conn *costandusagereportservice.CostandUsageReportService, reportName string) (*costandusagereportservice.ReportDefinition, error) {
	params := &costandusagereportservice.DescribeReportDefinitionsInput{}

	log.Printf("[DEBUG] Reading CurReportDefinition: %s", reportName)

	var matchingReportDefinition *costandusagereportservice.ReportDefinition
	err := conn.DescribeReportDefinitionsPages(params, func(resp *costandusagereportservice.DescribeReportDefinitionsOutput, isLast bool) bool {
		for _, reportDefinition := range resp.ReportDefinitions {
			if aws.StringValue(reportDefinition.ReportName) == reportName {
				matchingReportDefinition = reportDefinition
				return false
			}
		}
		return !isLast
	})
	if err != nil {
		return nil, err
	}
	return matchingReportDefinition, nil
}
```